controllers/drplacementcontrol_controller.go

```diff
@@ -2422,3 +2447,173 @@ func (r *DRPlacementControlReconciler) determineDRPCState(
 
 	return AllowFailover, msg, nil
 }
+
+// ensureVRGsManagedByDRPC ensures that VRGs reported by ManagedClusterView are managed by the current instance of
+// DRPC. This is done using the DRPC UID annotation on the viewed VRG matching the current DRPC UID and if not
+// creating or updating the exisiting ManifestWork for the VRG.
+// Returns a bool indicating true if VRGs are managed by the current DRPC resource
+func ensureVRGsManagedByDRPC(
+	log logr.Logger,
+	mwu rmnutil.MWUtil,
+	vrgs map[string]*rmn.VolumeReplicationGroup,
+	drpc *rmn.DRPlacementControl,
+	vrgNamespace string,
+) bool {
+	ensured := true
+
+	for cluster, viewVRG := range vrgs {
+		if viewVRG == nil {
+			continue
+		}
+
+		if rmnutil.ResourceIsDeleted(viewVRG) {
+			log.Info("VRG reported by view undergoing deletion, during adoption")
+
+			continue
+		}
+
+		if viewVRG.GetAnnotations() != nil {
+			if v, ok := viewVRG.GetAnnotations()[DRPCUIDAnnotation]; ok && v == string(drpc.UID) {
+				continue
+			}
+		}
+
+		adopted := adoptVRG(log, mwu, viewVRG, cluster, drpc, vrgNamespace)
+
+		ensured = ensured && adopted
+	}
+
+	return ensured
+}
+
+// adoptVRG creates or updates the VRG ManifestWork to ensure that the current DRPC is managing the VRG resource
+// Returns a bool indicating if adoption was completed (which is mostly false except when VRG MW is deleted)
+func adoptVRG(
+	log logr.Logger,
+	mwu rmnutil.MWUtil,
+	viewVRG *rmn.VolumeReplicationGroup,
+	cluster string,
+	drpc *rmn.DRPlacementControl,
+	vrgNamespace string,
+) bool {
+	adopted := true
+
+	mw, err := mwu.FindManifestWorkByType(rmnutil.MWTypeVRG, cluster)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Info("error fetching VRG ManifestWork during adoption", "error", err)
+
+			return !adopted
+		}
+
+		adoptOrphanVRG(log, mwu, viewVRG, cluster, drpc, vrgNamespace)
+
+		return !adopted
+	}
+
+	if rmnutil.ResourceIsDeleted(mw) {
+		log.Info("VRG ManifestWork found deleted during adoption")
+
+		return adopted
+	}
+
+	vrg, err := rmnutil.ExtractVRGFromManifestWork(mw)
+	if err != nil {
+		log.Info("error extracting VRG from ManifestWork during adoption", "error", err)
+
+		return !adopted
+	}
+
+	// NOTE: upgrade use case, to add DRPC UID for existing VRG MW
+	adoptExistingVRGManifestWork(log, mwu, vrg, cluster, drpc, vrgNamespace)
+
+	return !adopted
```

Comment:
Again, why `!adopted`? Aren't we adopting in this function, and the functions seems to be doing what is supposed to do? So maybe change this to:
```
return adoptExistingVRGManifestWork(log, mwu, vrg, cluster, drpc, vrgNamespace)
```

Reply:
For both this and `adoptOrphanVRG` the update of MW is really only starting the adoption, the adoption is complete once MVC reports that the resource has been applied on the ManagedCluster and reports the same UID.

This avoids cases where we proceed with the reconciliation, and overwrite the MW or delete it before it got a chance to be applied to the ManagedCluster.

The return of `adopted` bool hence is to address the one fake/corner case at 2517 (VRG deleted hence assume adopted). The alternate to that was return !adopted and wait for the VRG to be deleted on the cluster, but we really do not know if deletion is waiting for other aspects to be cleaned up and hence need to make reconciliation progress. So had to (chose to) return that deleted MW is assumed to be adopted.